File: leetcode/problems/152/152.cc
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int s = nums[0];
        int imax = s;
        int imin = s;
        for(int i = 1; i < nums.size(); i++){
            
            if(nums[i] < 0){
                swap(imax, imin);
            }
            
            imax = max(nums[i], nums[i] * imax);
            imin = min(nums[i], nums[i] * imin);
            
            s = max(s, imax);
        }
        return s;
    }
};

File: leetcode/problems/152/152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums)==0:
            return 0
        record=[nums[0],nums[0]]
        res=nums[0]
        
        for v in nums[1:]:
            record=[max(v,v*record[0],v*record[1]),min(v,v*record[0],v*record[1])]
            res=max(res,record[0])
            
        return res